File: app/src/main/java/autogravity/com/mvvmdemo/domain/Response.java
package autogravity.com.mvvmdemo.domain;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

public class Response<T> {
    @NonNull public final Status status;
    @Nullable public final T data;
    @Nullable public final String message;

    private Response(@NonNull Status status, @Nullable T data, @Nullable String message) {
        this.status = status;
        this.data = data;
        this.message = message;
    }

    public static <T> Response<T> success(@NonNull T data) {
        return new Response<>(Status.SUCCESS, data, null);
    }

    public static <T> Response<T> error(String msg, @Nullable T data) {
        return new Response<>(Status.ERROR, data, msg);
    }

    public static <T> Response<T> loading(@Nullable T data) {
        return new Response<>(Status.LOADING, data, null);
    }
}

File: app/src/main/java/autogravity/com/mvvmdemo/domain/Response.kt
package autogravity.com.mvvmdemo.domain

class Response<T> private constructor(val status: Status, val data: T?, val throwable: Throwable?) {
    companion object {

        fun <T> loading(): Response<T> {
            return Response(Status.LOADING, null, null)
        }

        fun <T> success(data: T): Response<T> {
            return Response(Status.SUCCESS, data, null)
        }

        fun <T> error(t: Throwable, data: T?): Response<T> {
            return Response(Status.ERROR, data, t)
        }
    }
}
